File: 53-maximum-subarray/maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        auto sum = 0, curr = 0;
        int n = nums.size();
        bool positive = false;
        for(int i=0; i<n; i++){
            if(nums[i] >= 0)
                positive = true;
            curr += nums[i];
            sum = max(sum, curr);
            if(curr < 0)
                curr = 0;
        }
        if(!positive)   return *max_element(nums.begin(), nums.end());
        return sum;
    }
};

File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int sum = nums[0], maxi = Integer.MIN_VALUE;
        for(int i=1; i<n; i++){
            maxi = Math.max(maxi, sum);
            if((sum + nums[i]) > nums[i])
                sum += nums[i];
            else
                sum = nums[i];
        }
        maxi = Math.max(maxi, sum);
        return maxi;
    }
}